src/include/simeng/memory/Mem.hh

```diff
@@ -1,74 +1,120 @@
 #pragma once
 
+#include <functional>
 #include <memory>
 
 #include "simeng/OS/Process.hh"
 
 namespace simeng {
 namespace memory {
-// This is a very basic implementation of Memory in SimEng, it has been kept
-// simple just to get dynamic linking and multicore simulation working. It is
-// very similar to the previous implementation but unifies both instruction and
-// data memory. Whereas, previously two different copies of process memory were
-// for instruction and data memory interfaces.
-
-/* Enum for classifying data access. */
-enum DataPacketAccessType { READ, WRITE };
-
-/* A data packet represents access to memory. This has named and cnstruct such
- * that future improvements to memory system can be facilitated
- */
-struct DataPacket {
-  static uint64_t pktIdCtr;
-  uint64_t id;
-  DataPacketAccessType type;
-  DataPacket(DataPacketAccessType accType);
-};
+/* Enum for classifying DataPacket type. */
+enum DataPacketType {
+  /** This type signifies a read request to memory. */
+  READ_REQUEST,
 
-/* Response to a read packed. */
-struct ReadRespPacket : public DataPacket {
-  uint64_t reqId;
-  size_t bytesRead;
-  char* data;
-  ReadRespPacket(uint64_t req_id, size_t bytes_read, char* dt);
-};
+  /** This type signifies a write request to memory.*/
+  WRITE_REQUEST,
 
-/* ReadPacket represents a read access to data packet. */
-struct ReadPacket : public DataPacket {
-  uint64_t address;
-  size_t size;
+  /** This type signifies the response to a read request. */
+  READ_RESPONSE,
 
-  ReadPacket(uint64_t addr, size_t sz);
-  ReadRespPacket* makeResponse(uint64_t bytesRead, char* data);
-};
+  /** This type signifies the response to a write request. */
+  WRITE_RESPONSE,
 
-/* Response to a write packed. */
-struct WriteRespPacket : public DataPacket {
-  uint64_t reqId;
-  size_t bytesWritten;
-  WriteRespPacket(uint64_t req_id, size_t bytes_written);
+  /** This type signifies a faulty or empty request of no type. */
+  NONE
 };
 
-/* WritePacket represents a write access to a data packet. */
-struct WritePacket : public DataPacket {
-  uint64_t address;
-  size_t size;
-  const char* data;
+/* A data packet struct contains information which enables access to the
+ * simulation memory. */
+struct DataPacket {
```

Comment:
Given this is now more complex, is it worth defining as a class rather than a struct (functionally equivalent, but maybe sensible for more complex structs as has been done in other cases?)

Reply:
I don't think it is, it is still just used to group data and transport it. So it makes sense to keep it as a struct. 
Look at [these](https://google.github.io/styleguide/cppguide.html#Structs_vs._Classes) guidelines from